File: typescript/doubleLinkedList.ts
type Node<T> = {
	value: T;
	prev?: Node<T>;
	next?: Node<T>;
};

export default class DoublyLinkedList<T> {
	public length: number;
	private head?: Node<T>;
	private tail?: Node<T>;

	constructor() {
		this.length = 0;
		this.head = undefined;
		this.tail = undefined;
	}

	prepend(item: T): void {
		const node = { value: item } as Node<T>;
		this.length++;

		if (!this.head) {
			this.head = this.tail = node;
			return;
		}

		node.next = this.head;
		this.head.prev = node;
		this.head = node;
	}
	insertAt(item: T, idx: number): void {
		if (idx > this.length) {
			throw new Error('Index out of bounds');
		}

		if (idx == this.length) {
			this.append(item);
			return;
		} else if (idx == 0) {
			this.prepend(item);
			return;
		}

		this.length++;

		const current = this.getAt(idx) as Node<T>;

		const node = { value: item } as Node<T>;

		node.next = current;
		node.prev = current.prev;
		current.prev = node;

		if (node.prev) {
			node.prev.next = node;
		}
	}
	append(item: T): void {
		this.length++;
		const node = { value: item } as Node<T>;
		if (!this.tail) {
			this.head = this.tail = node;
		}

		this.tail.next = node;
		node.prev = this.tail;
		this.tail = node;
	}
	remove(item: T): T | undefined {
		let current = this.head;

		for (let i = 0; current && i < this.length; i++) {
			if (current.value == item) {
				break;
			}

			current = current.next;
		}

		if (!current) {
			return;
		}

		this.removeNode(current);
	}
	get(idx: number): T | undefined {
		const node = this.getAt(idx);
		return node?.value;
	}
	removeAt(idx: number): T | undefined {
		const node = this.getAt(idx);

		if (!node) {
			return undefined;
		}

		return this.removeNode(node);
	}

	private removeNode(node: Node<T>): T | undefined {
		this.length--;

		if (this.length == 0) {
			const out = this.head?.value;
			this.head = this.tail = undefined;
			return out;
		}

		if (node.prev) {
			node.prev.next = node.next;
		}

		if (node.next) {
			node.next.prev = node.prev;
		}

		if (node == this.head) {
			this.head = node.next;
		}

		if (node == this.tail) {
			this.tail = node.prev;
		}

		node.next = node.prev = undefined;

		return node.value;
	}

	private getAt(idx: number): Node<T> | undefined {
		let current = this.head;
		for (let i = 0; current && i < idx; i++) {
			current = current.next;
		}

		return current;
	}
}

const list = new DoublyLinkedList<number>();
list.append(1);
list.append(2);
list.append(3);
list.removeAt(1);
console.log(list.get(0));
console.log(list.get(1));

File: python/double_linked_list.py
from typing import Optional

class Node:
    def __init__(self, value):
        self.value = value
        self.prev: Optional[Node] = None
        self.next: Optional[Node] = None
        
class Double_Linked_List:
    def __init__(self) -> None:
        self.head = None
        self.tail = None
        self.length = 0
        
    def __get_at(self, index):
        current = self.head
        for _n in range(index):
            if current == None:
                break
            current = current.next
            
        return current
    
    def __remove_node(self, node: Node):
        self.length -= 1
        if self.length == 0 and self.head:
            out = self.head.value
            self.head = self.tail = None
            return out
        
        if node.prev:
            node.prev.next = node.next
            
        if node.next:
            node.next.prev = node.prev
            
        if node == self.head:
            self.head = node.next
            
        if node == self.tail:
            self.tail = node.prev
            
        node.prev = node.next = None
        
        return node.value
        
    def prepend(self, item):
        new_node = Node(item)
        self.length += 1
        
        if self.head is None:
            self.head = self.tail = new_node
            
        new_node.next = self.head
        self.head.prev = new_node
        self.head = new_node
        
    def append(self, item):
        new_node = Node(item)
        self.length += 1
        
        if self.tail is None:
            self.head = self.tail = new_node
            
        new_node.prev = self.tail
        self.tail.next = new_node
        self.tail = new_node
            
    def insert_at(self, item, index):
        if index > self.length:
            raise IndexError
        
        if index == self.length:
            self.append(item)
            return 
        elif index == 0:
            self.prepend(item)
            return
        
        self.length += 1
        
        current = self.__get_at(index)
        node = Node(item)
        
        node.next = current
        if current:
            node.prev = current.prev
            current.prev = node
        
        if node.prev:
            node.prev.next = node
            
    def remove(self, item):
        current = self.head
  
        for _n in range(self.length):
            if current and current.value == item:
                break
            if current:
                current = current.next
                
        if (current == None):
            return
		
  
        self.__remove_node(current)
	
    def get(self, idx):
        node = self.__get_at(idx)
        if node:
            return node.value
        else:
            return None
    
	
    def removeAt(self, idx):
        node = self.__get_at(idx)

        if (node == None):
            return None
        
        return self.__remove_node(node);
	
 
my_doble_linked_list = Double_Linked_List()
my_doble_linked_list.append(1);
my_doble_linked_list.append(2);
my_doble_linked_list.append(3);
my_doble_linked_list.removeAt(1);
print(my_doble_linked_list.get(0));
print(my_doble_linked_list.get(1));
